main/java/mjc/Runner.java

```diff
@@ -0,0 +1,155 @@
+
+            printMap(weekdayPurchaseEnumMap, "Weekday => List of purchases");
+
+            System.out.println("Total cost of purchases in a single day");
+            for (Map.Entry<Weekday, List<AbstractPurchase>> entry : weekdayPurchaseEnumMap.entrySet()) {
+                System.out.println(entry.getKey() + " " + getTotalCost(entry.getValue()));
+            }
+
+            findAndPrintMapValue(weekdayPurchaseEnumMap, Weekday.MONDAY, "\nFind purchases on MONDAY");
+            removeMapEntries(weekdayPurchaseEnumMap, new EntryChecker<>() {
+                @Override
+                public boolean check(Map.Entry<Weekday, List<AbstractPurchase>> entry) {
+                    return entry.getValue().removeIf(purchase -> purchase.getProduct().getName().equals("milk"));
+                }
+            });
+            printMap(weekdayPurchaseEnumMap, "\nRemoved entries where name is 'milk'");
+        } catch (IOException e) {
+            System.err.println("No source file found.");
+        }
+    }
+
+    public interface EntryChecker<K, V> {
+        boolean check(Map.Entry<K, V> entry);
+    }
+
+    private static <K, V> void removeMapEntries(Map<K, V> map, EntryChecker<K, V> checker) {
+        map.entrySet().removeIf(checker::check);
+    }
+
+    private static void putFirstDay(Map<AbstractPurchase, Weekday> map, Weekday day, AbstractPurchase purchase) {
+        if (!map.containsKey(purchase)) {
+            map.put(purchase, day);
+        }
+    }
+
+    private static void putLastDay(Map<AbstractPurchase, Weekday> map, Weekday day, AbstractPurchase purchase) {
+        map.put(purchase, day);
+    }
+
+    private static void putEnumMapPurchase(Map<Weekday, List<AbstractPurchase>> map, Weekday day, AbstractPurchase purchase) {
+        List<AbstractPurchase> purchases = map.get(day);
+        if (purchases == null) {
+            map.put(day, purchases = new ArrayList<>());
+        }
+        purchases.add(purchase);
+    }
+
+    private static <K, V> void printMap(Map<K, V> map, String header) {
+        System.out.println(header);
+        for (Map.Entry<K, V> entry : map.entrySet()) {
+            System.out.println(entry.getKey() + "=>" + entry.getValue().toString());
+        }
+        System.out.println();
+    }
+
+    private static <K, V> void findAndPrintMapValue(Map<K, V> map, K searchKey, String header) {
+        System.out.println(header);
+        V value = map.get(searchKey);
+        if (value == null) {
+            System.out.println("No value found.");
+        } else {
+        System.out.println(value);
+        }
+    }
+
+    private static Euro getTotalCost(List<?> purchases) {
+        Euro totalCost = new Euro(0);
+        for (Object purchase : purchases) {
+            if (purchase instanceof AbstractPurchase p) {
+                totalCost = totalCost.add(p.getCost());
+            }
+        }

```

Comment:
i'm sure you know that this implementation is bad. 
you could ask me. 
because it's the only essential drawback, i've shown the right version without any prompts ) 
```suggestion
    private static Euro getTotalCost(List<? extends AbstractPurchase> purchases) {
        Euro totalCost = new Euro(0);
        for (AbstractPurchase purchase : purchases) {
            totalCost = totalCost.add(p.getCost());
        }
```
to bound type arg is a common practice.